### PR TITLE
proto: apply timeout to TLS/QUIC/H3 handshake phase

### PR DESCRIPTION
**Pull request by japaric**
a timeout is applied to [the TCP connection phase](https://github.com/hickory-dns/hickory-dns/blob/d40f11fb7307f9fb9707750dea6619220802ba98/crates/proto/src/runtime.rs#L182) but no timeout is applied to TLS handshake. this results in clients "hanging" (blocking forever) when trying to establish a TLS connection to a DNS server.

this PR applies a timeout to the TLS handshake to prevent the problem.

an E2E test has been added to serve as a regression test.

### CODE REVIEW

**Comment by djc** [comment]
This seems to affect DoH2 only. Do we need similar fixes for DoT/DoQ/DoH3?

**Comment by japaric** [comment]
>  Do we need similar fixes for DoT/DoQ/DoH3? And what about the native-tls based TLS support?

I have added E2E tests for DoT, DoQ and DoH3 as they are all affected. I have also added a handshake timeout to DoT (fixing that test) but DoQ and DoH3 still need fixing.

with respect to native-tls, the `dns` binary does not expose that as a Cargo feature and always uses rustls. one option would be add a new Cargo feature to `dns` to use native-tls instead of rustls and then build `dns` with native-tls enabled in the hickory.Dockerfile and have it as a separate binary, eg `dns-native-tls`. then a new test case can be added to use `dns-native-tls` instead of `dns`

File: crates/proto/src/runtime.rs [hunk 1]
```diff
@@ -122,6 +122,9 @@ mod tokio_runtime {
     use tokio::task::JoinSet;
     use tokio::time::timeout;
 
+    pub const TLS_TIMEOUT: Duration = Duration::from_secs(5);
+    pub const TCP_TIMEOUT: Duration = Duration::from_secs(5);
```
Comment (djc) [inline crates/proto/src/runtime.rs]: Nit: keep these at the bottom of the module, please.

Not sure about where we should keep these timeouts. They're arguably not really a related to the runtime, that's merely the lowest level at which we can enforce the TCP timeout (but not the TLS timeout). Maybe we should move these into `src/xfer/mod.rs` or something?

@marcus0x62 thoughts?

File: crates/proto/src/h2/h2_client_stream.rs [hunk 2]
```diff
@@ -466,7 +475,10 @@ where
                     match ServerName::try_from(&*tls.dns_name) {
                         Ok(dns_name) => {
                             let tls = TlsConnector::from(tls.client_config);
-                            let tls = tls.connect(dns_name.to_owned(), AsyncIoStdAsTokio(tcp));
+                            let tls = Box::pin(time::timeout(
```
Comment (djc) [inline crates/proto/src/h2/h2_client_stream.rs:477]: Nit: I'd prefer to use `timeout()` directly, I don't think the `time::` qualification adds anything other than clutter.

File: crates/proto/src/runtime.rs [hunk 3]
```diff
@@ -178,7 +181,7 @@ mod tokio_runtime {
 
                 socket.set_nodelay(true)?;
                 let future = socket.connect(server_addr);
-                let wait_for = wait_for.unwrap_or_else(|| Duration::from_secs(5));
+                let wait_for = wait_for.unwrap_or(TCP_TIMEOUT);
```
Comment (djc) [inline crates/proto/src/runtime.rs:189]: Nit: please make this a separate commit.

File: crates/proto/src/runtime.rs [hunk 4]
```diff
@@ -122,6 +122,7 @@ mod tokio_runtime {
     use tokio::task::JoinSet;
     use tokio::time::timeout;
 
+    #[cfg(any(feature = "dns-over-rustls", feature = "dns-over-https-rustls"))]
```
Comment (djc) [inline crates/proto/src/runtime.rs:125]: Squash this into the originating commit, please?

**Review by djc — COMMENTED** [review]
Mostly looks good -- not sure where we should manage the default timeouts.

File: crates/proto/src/runtime.rs (hunk 1 above)
Comment (marcus0x62) [reply crates/proto/src/runtime.rs]: I think for now, these static timeouts should go in mod.rs as @djc suggests.  In the long-term, I think we'll end up with more timeouts, and will need to make the timeouts configurable; at that point, we'll probably want a struct or something to put these in and pass them around, but that seems like overkill at the moment.

File: crates/proto/src/runtime.rs [hunk 5]
```diff
@@ -122,6 +122,14 @@ mod tokio_runtime {
...
+    pub const TLS_TIMEOUT: Duration = Duration::from_secs(5);
```
Comment (marcus0x62) [inline crates/proto/src/runtime.rs:126]: I'd suggest calling these <PROTO>_HANDSHAKE_TIMEOUT

Comment (pvdrz) [reply crates/proto/src/runtime.rs]: But there are several of these, how would you name them? TLS_PROTO_HANDSHAKE_TIMEOUT and so on?
